js/src/table.ts

```diff
@@ -0,0 +1,272 @@
+            const columns = this.parent.batches[batch];
+            const predicate = this.predicate.bind(columns);
+
+            // yield all indices
+            for (let idx = -1; ++idx < length;) {
+                if (predicate(idx, columns)) { next(idx, columns); }
+            }
+        }
+    }
+
+    count(): number {
+        // inlined version of this:
+        // let sum = 0;
+        // this.parent.scan((idx, columns) => {
+        //     if (this.predicate(idx, columns)) ++sum;
+        // });
+        // return sum;
+        let sum = 0;
+        for (let batch = -1; ++batch < this.parent.lengths.length;) {
+            const length = this.parent.lengths[batch];
+
+            // load batches
+            const columns = this.parent.batches[batch];
+            const predicate = this.predicate.bind(columns);
+
+            // yield all indices
+            for (let idx = -1; ++idx < length;) {
+                if (predicate(idx, columns)) { ++sum; }
+            }
+        }
+        return sum;
+    }
+
+    filter(predicate: Predicate): DataFrame {
+        return new FilteredDataFrame(
+            this.parent,
+            this.predicate.and(predicate)
+        );
+    }
+
+    countBy(count_by: (Col|string)): CountByResult {
+        if (!(count_by instanceof Col)) {
+            count_by = new Col(count_by);
+        }
+
+        // Assume that all dictionary batches are deltas, which means that the
+        // last record batch has the most complete dictionary
+        count_by.bind(this.parent.batches[this.parent.batches.length - 1]);
+        if (!(count_by.vector instanceof DictionaryVector)) {
+            throw new Error('countBy currently only supports dictionary-encoded columns');
+        }
+
+        const data: Vector = (count_by.vector as DictionaryVector<any>).data;
+        // TODO: Adjust array byte width based on overall length
+        // (e.g. if this.length <= 255 use Uint8Array, etc...)
+        const counts: Uint32Array = new Uint32Array(data.length);
+
+        for (let batch = -1; ++batch < this.parent.lengths.length;) {
+            const length = this.parent.lengths[batch];
+
+            // load batches
+            const columns = this.parent.batches[batch];
+            const predicate = this.predicate.bind(columns);
+            count_by.bind(columns);
+            const keys: Vector = (count_by.vector as DictionaryVector<any>).keys;
+
+            // yield all indices
+            for (let idx = -1; ++idx < length;) {
+                let key = keys.get(idx);
+                if (key !== null && predicate(idx, columns)) { counts[key]++; }
+            }
+        }
+
+        return new CountByResult(data, new Uint32Vector({data: counts}));
+    }
+}
+
+export class CountByResult extends Table implements DataFrame {
+    constructor(readonly values: Vector, readonly counts: Vector<number|null>) {
+        super({batches: [[values, counts]]});
+    }
+
+    asJSON(): Object {
```

Comment:
@TheNeuralBit what about renaming this to `toJSON()`, so it also becomes the [default serialization behavior](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) if an instance is serialized via JSON.stringify()?

Reply:
ah nice I didn't know about `toJSON`, I'll make that change